### ISSUE TITLE
blocking or non-blocking?

### ISSUE THREAD

**Issue by yingqiaomxi** [ef-labs/vertx-jersey#56]
I am interested in using Jersey to handle RESTful requests in vertx. However, I found that the JerseyServer in the vertx-jersey handles requests in a blocking manner. For example, I added a executeBlocking block in a resource method, and made the running of a request stop there. Then, I sent the second request to the same endpoint. However, the second request was not forwarded to the same method at that time. It was forwarded to the method only after the method sent out the AsynResponse.  

I tried the same experiment in when.java example. I got the same results. 

I wonder if there is a way that I could use to solve this problem?

**Comment by adrianluisgonzalez** [comment]
Not sure I follow, can you provide a code sample?

**Comment by yingqiaomxi** [comment]
Yes.

I have a resource as follows: 
@Path("/")
public class HelloWorldEndpoint {

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public void getString(@Suspended AsyncResponse asyncResponse,
                          @Context Vertx aVertx) throws IOException {
        Future lFuture = Future.future();
        
        aVertx.executeBlocking(aFuture->{
             aFuture.complete();
        }, false,
                aResult->{
                     lFuture.complete();
                });

        lFuture.setHandler(aResult->{
            IDOut aID = new IDOut(UUID.randomUUID().toString());
            asyncResponse.resume(aID);
        });
    }

    @Path("/second")
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public void getSecondString(@Suspended AsyncResponse asyncResponse,
            @Context Vertx aVertx) throws IOException {

        Future lFuture = Future.future();

        aVertx.executeBlocking(aFuture->{
             aFuture.complete();
                }, false,
                aResult->{
                    lFuture.complete();
                });


        lFuture.setHandler(aResult->{
            IDOut aID = new IDOut(UUID.randomUUID().toString());
            asyncResponse.resume(aID);
        });
    }
}

I have JerseyServer running in a bedug mode, and I set the break points in the methods of resource on Future lFuture = Future.future(), aFuture.complete() and asynResponse.resume() statements, respectively. Through a browser, I sent a GET request to "/". The server stopped at first breakpoint, and then, I made it stopped at aFuture.complete(). After that, I sent the second request to "/", and the server still stopped at the second breakpoint. In the case of non-blocking, there should be a running thread waiting at the first breakpoint right at this time, right? 

I saw a running thread stopping at the first breakpoint after the asyncResponse.resume() statement was executed. 

So, is this a case of blocking?

**Comment by yingqiaomxi** [comment]
I added an executeBlocking in the getString of WhenJerseyServerResouce. The requests are run in non-blocking. 
The code is here:
@Path("when")
public class WhenJerseyServerResource {

    private final When when;

    @Inject
    public WhenJerseyServerResource(When when) {
        this.when = when;
    }

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public void getString(@Suspended AsyncResponse asyncResponse,
                          @Context Vertx aVertx) throws IOException {

        Deferred<String> d = when.defer();

        d.getPromise()
                .then(str -> {
                    asyncResponse.resume(str);
                    return null;
                })
                .otherwise(t -> {
                    asyncResponse.resume(new WebApplicationException(t));
                    return null;
                });

        aVertx.executeBlocking(aFuture->{
            aVertx.setTimer(5000, id->{
                String currentTime = new SimpleDateFormat("HH:mm:ss").format(new Date());
               aFuture.complete(currentTime);
            });
        
        }, false, aResult->{
            d.resolve("when.java!" + aResult.result().toString() +"\n");
        });
    }
}

So, non-blocking only works with vertx-jersey and vertx-when? Does it work with vertx-jersey only?

**Comment by adrianluisgonzalez** [comment]
I'm not sure what it is you are trying to demonstrate, but I will try to explain the `vertx-jersey` threading:

The `DefaultJerseyHandler` runs on the vert.x event loop.  You obviously should not write blocking code on this thread.  The expectation is you either make non-blocking calls (db, http, event bus, etc.) or if there is blocking code you use `Vertx#executeBlocking` as you did.

I have adapted your sample code a little to log with thread names and simulated blocking:

```
package com.eflabs.vertx.resources;

import io.vertx.core.Future;
import io.vertx.core.Vertx;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;

import javax.inject.Singleton;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.container.AsyncResponse;
import javax.ws.rs.container.Suspended;
import javax.ws.rs.core.Context;
import java.util.concurrent.atomic.AtomicInteger;

@Singleton
@Path("/")
public class HelloWorldEndpoint {

    private static final AtomicInteger counter = new AtomicInteger();
    private static final Logger logger = LoggerFactory.getLogger(HelloWorldEndpoint.class);

    @GET
    public void getString(
            @Context Vertx vertx,
            @Suspended AsyncResponse asyncResponse) {

        int id = counter.incrementAndGet();
        log("Starting", id);

        Future future = Future.future()
                .setHandler(aResult -> {
                    log("Resuming response", id);
                    asyncResponse.resume(id);
                });

        vertx.executeBlocking(
                blockingFuture -> {
                    try {
                        // Simulate blocking code
                        Thread.sleep(5000);
                        log("Blocking complete", id);
                        blockingFuture.complete();
                    } catch (InterruptedException e) {
                        throw new RuntimeException(e);
                    }
                }, false,
                result -> {
                    log("Future complete", id);
                    future.complete();
                });

    }

    private void log(String msg, int count) {
        logger.info(msg + " [" + count + " - " + Thread.currentThread().getName() + "]");
    }
}
```

Sending 3 http requests in parallel logs the following:

```
Jun 05, 2017 4:13:22 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Starting [1 - vert.x-eventloop-thread-0]
Jun 05, 2017 4:13:22 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Starting [2 - vert.x-eventloop-thread-0]
Jun 05, 2017 4:13:22 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Starting [3 - vert.x-eventloop-thread-0]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Blocking complete [1 - vert.x-worker-thread-1]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Future complete [1 - vert.x-eventloop-thread-0]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Resuming response [1 - vert.x-eventloop-thread-0]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Blocking complete [2 - vert.x-worker-thread-2]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Future complete [2 - vert.x-eventloop-thread-0]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Resuming response [2 - vert.x-eventloop-thread-0]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Blocking complete [3 - vert.x-worker-thread-3]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Future complete [3 - vert.x-eventloop-thread-0]
Jun 05, 2017 4:13:27 PM com.eflabs.vertx.resources.HelloWorldEndpoint
INFO: Resuming response [3 - vert.x-eventloop-thread-0]
```

Hopefully you can agree this demonstrates `vertx-jersey` is non-blocking.

**Comment by yingqiaomxi** [comment]
Many thanks, @adrianluisgonzalez! 
It seems that it is not necessary to use vertx-when for non-blocking, right?

**Comment by adrianluisgonzalez** [comment]
Correct, `vertx-jersey` is non-blocking on its own.  You should use `vertx-when` if you prefer the Promises programming style over callbacks or rx.

**Comment by yingqiaomxi** [comment]
Great!
Many thanks, @adrianluisgonzalez!

**Comment by yingqiaomxi** [comment]
Many thanks for your help @adrianluisgonzalez! I am using vertx-jersey for my project, and it works very well. 
I am using vertx.executeBlocking in resource methods now since these methods would access the data stored in a database. Jersey has @ManagedAsyn annotation for running resource method asynchronously. Between executeBlocking and @ManagedAsyn, which is the better way to use with vertx? If I use both of them, is there any defect this approach would induce?

**Comment by adrianluisgonzalez** [comment]
It's your decisions, but I would stick with the vertx event loop.  If Jersey starts creating threads for you then you lose the benefits of the vertx event loop and are back to the overhead of thread context switching.

**Comment by yingqiaomxi** [comment]
Good point, and thanks!
